heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,290 @@
+
+    if (request.getInstancesCount() > 0) {
+      instances.addAll(request.getInstancesList());
+    } // empty list means all instances
+
+    return request1;
+  }
+
+  // compatible with com.twitter.heron.proto.tmaster.TopologyMaster.ExceptionLogResponse
+  public static TopologyMaster.ExceptionLogResponse Convert(ExceptionResponse response) {
+    TopologyMaster.ExceptionLogResponse.Builder builder =
+        TopologyMaster.ExceptionLogResponse.newBuilder();
+    // default OK if we have response to build already
+    builder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+
+    for (ExceptionDatapoint e : response.exceptionDatapointList) {
+      TopologyMaster.TmasterExceptionLog.Builder builder1 =
+          TopologyMaster.TmasterExceptionLog.newBuilder();
+      // ExceptionDatapoint
+      builder1.setComponentName(e.componentName);
+      builder1.setHostname(e.hostname);
+      builder1.setInstanceId(e.instanceId);
+      // ExceptionData
+      builder1.setStacktrace(e.stacktrace);
+      builder1.setLasttime(e.lasttime);
+      builder1.setFirsttime(e.firsttime);
+      builder1.setCount(e.count);
+      builder1.setLogging(e.logging);
+
+      builder.addExceptions(builder1);
+    }
+
+    return builder.build();
+  }
+
+
+  public static class ExceptionRequest {
+    public Map<String, Set<String>> componentNameInstanceId;
+  }
+
+  public static class ExceptionResponse {
+    public List<ExceptionDatapoint> exceptionDatapointList;
+  }
+
+
+  public static class MetricTimeRangeValue {
+    public long startTime;
+    public long endTime;
+    public String value;
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("[")
+          .append(startTime).append("-").append(endTime)
+          .append(":")
+          .append(value)
+          .append("]");
+      return sb.toString();
+    }
+  }
+
+  public static class MetricDatum {
+    public String componentName;
+    public String instanceId;
+    public String metricName;
+    public List<MetricTimeRangeValue> metricValue;
+  }
+
+  public static class MetricRequest {
+    // The instance ids to get the stats from
+    // If nothing is specified, we will get from
+    // all the instances of the component name
+    public Map<String, Set<String>> componentNameInstanceId;
+    // What set of metrics you are interested in
+    // Example is __emit-count/default
+    public Set<String> metricNames;
+    // what timeframe data in milliseconds
+    // -1/0 means everything
+    // What is the time interval that you want the metrics for
+    // Clients can specify one in many ways.
+    // 1. interval. Essentially seconds worth of metrics before the current time
+    // 2. explicit_interval. An explicit interval
+    public long startTime;
+    public long endTime;
+    // Do you want metrics broken down on a per minute basis?
+    public int minutely;
```

Comment:
boolean isMinutely. Or if we can support multiple granularities, better to make the interval an enum.